R/scidb.R

```
@@ -0,0 +1,1215 @@
+      stopifnot("invalid argument specified for get_entity()" = missing(...))
+      keep_hidden_fields <- identical(TRUE, keep_hidden_fields)
+      stopifnot(
+        "character vector expected for 'keep_fields'" =
+          is.null(keep_fields) || (is.character(keep_fields) && length(keep_fields) > 0)
+      )
+      stopifnot(
+        "character vector expected for 'drop_fields'" =
+          is.null(drop_fields) || (is.character(drop_fields) && length(drop_fields) > 0)
+      )
+      full_arrayname <- full_arrayname(
+        pkg_schema = self$rvlEnv,
+        entitynm = entitynm,
+        con = self
+      )
+
+      # Figure out our hidden-field-situation
+      fields_to_keep <- c(names(get_fields(self$rvlEnv, entitynm)))
+      hidden_fields <- setNames(grep("^_", fields_to_keep, value = TRUE), NULL)
+      if (!keep_hidden_fields && length(hidden_fields) > 0) {
+        drop_fields <- c(drop_fields, hidden_fields)
+      }
+
+      # Setup fields filter, if specified
+      filter_fields <- FALSE
+      if (!is.null(keep_fields)) {
+        fields_to_keep <- intersect(fields_to_keep, keep_fields)
+        filter_fields <- TRUE
+      }
+      if (!is.null(drop_fields)) {
+        fields_to_keep <- setdiff(fields_to_keep, drop_fields)
+        filter_fields <- TRUE
+      }
+
+      # Check for empty fields list
+      stopifnot("cannot select zero attributes" = length(fields_to_keep) > 0)
+
+      # Construct entity query and cache
+      cache_name <- glue_me("ENTITY: {entitynm}, {fields_to_keep}")
+      if (!private$aop_has_cache(cache_name)) {
+        scanned_entity <- scan_entity(
+          pkg_schema = self$rvlEnv,
+          entitynm = entitynm,
+          con = self
+        )
+        if (full_arrayname == scanned_entity) {
+          ret <- self$aop_connection$array(full_arrayname)
+        } else {
+          ret <- self$aop_connection$afl_expr(scanned_entity)
+        }
+        attr(ret, ATTR_FULL_ARRAYNAME) <- full_arrayname
```

complete this functionality and remove `get_array()` -- the (default) behavior on hiding fields should be implemented and described more clearly.